### PR TITLE
[FEATURE] SST-HT for Multi-target regression

### PR DESCRIPTION
**Pull request by smastelini**
Greetings @jacobmontiel,

The aim of this pull request is to add the Stacked Single-target Regression Hoeffding Tree (SST-HT) algorithm, which was proposed as part of my undergoing PhD [(paper available here)](https://arxiv.org/abs/1903.12483). Currently, the paper is under review and it uses the code that is added in this PR. 

Changes proposed in this pull request:

* Addition of SST-HT algorithm for MTR
* Addition of automated functionality tests for SST-HT
* Documentation about the method's usage

---

Checklist
- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] Tests are included for new functionality or updated accordingly.
- [x] Travis CI build passes with no errors.
- [x] Test Coverage is maintained (threshold is -0.2%).
- [x] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

### CODE REVIEW

**Comment by jacobmontiel** [comment]
Thanks for the submission @smastelini . Do you have an estimation about when the paper review will be completed?

There seems to be an error from an unrelated test file (`test_show_versions.py`), I will check into that as soon as I am free from other tasks. Any help is more than welcomed.

**Comment by jacobmontiel** [comment]
A quick look into the error took me here:

https://github.com/scikit-learn/scikit-learn/pull/14205

The error is triggered by a change in scikit-learn.

**Comment by smastelini** [comment]
Hi @jacobmontiel,

Thank your for your rapid response!

When the tests failed I wondered whether or not I had done something wrong :/.

Is there anything I could do to help tackle this problem?

Regarding the paper revision, last time I checked I was still under review. It is yet the first round of revisions. But I hope we will get a feedback soon!

**Comment by jacobmontiel** [comment]
The fix (a01e22e27b89929f2d37f8dea7196770a5518415) to the show_versions error is now available in the master branch. Please merge it and let me know if it is working.

**Comment by smastelini** [comment]
Thank you Jacob for the fix! Everything seems to be working fine right now!

**Comment by jacobmontiel** [comment]
@smastelini Travis failed for the last commit due to a problem in the Windows step on their side. I relaunched the job and it finished successfully.

Coverage has increased although there are still some unreached [lines](https://codecov.io/gh/scikit-multiflow/scikit-multiflow/pull/123/diff?src=pr&el=tree#diff-c3JjL3NrbXVsdGlmbG93L3RyZWVzL211bHRpX3RhcmdldF9yZWdyZXNzaW9uX2hvZWZmZGluZ190cmVlLnB5)

**Comment by smastelini** [comment]
Hello @jacobmontiel, I am currently working on that matter. Thank you for the feedback!

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+from skmultiflow.utils.utils import get_dimensions
+from skmultiflow.trees.intra_cluster_variance_reduction_split_criterion \
+     import IntraClusterVarianceReductionSplitCriterion
+
+
+_PERCEPTRON = 'perceptron'
+_ADAPTIVE = 'adaptive'
+
+
+class StackedSingleTargetRegressionHoeffdingTree(
```
Comment (jacobmontiel) [inline src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: There is currently no standard to differentiate between classifier and regressors which can be confusing. Maybe we can go with an explicit approach as in `sklearn`, for example in this case:

`StackedSingleTargetRegressionHoeffdingTree` -> `StackedSingleTargetHoeffdingTreeRegressor`

Comments?

File: tests/trees/test_multi_target_regression_hoeffding_tree.py [hunk 2]
```diff
@@ -6,16 +6,16 @@
...
                                  n_informative=15, random_state=1,
                                  n_targets=3)
     stream.prepare_for_use()
 
     learner = MultiTargetRegressionHoeffdingTree(leaf_prediction='mean')
 
     cnt = 0
-    max_samples = 500
-    wait_samples = 10
+    max_samples = 2000
```
[the remark this answers is not in the document]
Comment (smastelini) [reply tests/trees/test_multi_target_regression_hoeffding_tree.py]: Yes, the idea here is to allow the tree trying to split more times. Hence, reaching more parts of the code.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+            self.fMAE_P = 0.95 * self.fMAE_P + np.absolute(
+                normalized_target_value - normalized_base_pred[:-1]
+            )
+
+            self.fMAE_SP = 0.95 * self.fMAE_SP + np.absolute(
+                normalized_target_value - normalized_meta_pred
+            )
+
+    # ===========================================
+    # == Hoeffding Regression Tree implementation ===
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: Thank you for the observation, this gonna be corrected in the next commit.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 4]
```diff
@@ -0,0 +1,773 @@
...
+                            else:
+                                predictions[i, j] = pred_SP[j]
+
+        return predictions
+
+    def _attempt_to_split(self, node, parent, parent_idx: int):
+        """Attempt to split a node.
+
+        If there exists significant variance among the target space of the
+        seem examples:
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: That's right! I've corrected that.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py (hunk 1 above)
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: I think that is a good idea for standardization! Nonetheless, how would these changes impact in the remaining (and existing) methods documentation? Would these changes be gradually implemented?

If you agree, I will create a new commit changing the names in both the algorithm and the corresponding test files (just for SST-HT for now). Let me know what you think! And thank you Jacob for your patience and attention during this PR revision process!

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 5]
```diff
@@ -0,0 +1,773 @@
...
+        def _predict_meta(self, X):
+            return self.perceptron_weight[1] @ X
+
+    class InactiveLearningNodeAdaptive(InactiveLearningNodePerceptron):
+
+        def __init__(self, initial_class_observations, perceptron_weight=None,
+                     random_state=None):
+            """InactiveLearningNodeAdaptive class constructor
+
+            Parameters
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: The documentation was updated in the last commit, as requested.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 6]
```diff
@@ -0,0 +1,773 @@
...
+            # Ignore added bias term in the comparison
+            self.fMAE_P = 0.95 * self.fMAE_P + np.absolute(
+                normalized_target_value - normalized_base_pred[:-1]
+            )
+
+            self.fMAE_SP = 0.95 * self.fMAE_SP + np.absolute(
+                normalized_target_value - normalized_meta_pred
+            )
+
+    class InactiveLearningNodePerceptron(MultiTargetRegressionHoeffdingTree.
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: I've added documentation for this class in the last commit.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 7]
```diff
@@ -0,0 +1,773 @@
...
+
+    class LearningNodePerceptron(MultiTargetRegressionHoeffdingTree.
+                                 LearningNodePerceptron):
+
+        def __init__(self, initial_class_observations, perceptron_weight=None,
+                     random_state=None):
+            """
+            LearningNodePerceptron class constructor
+
+            Parameters
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: The documentation was improved in the last commit.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py [hunk 8]
```diff
@@ -0,0 +1,773 @@
...
+            else:
+                return self._observed_class_distribution[0]
+
+    class LearningNodeAdaptive(LearningNodePerceptron):
+
+        def __init__(self, initial_class_observations, perceptron_weight=None,
+                     random_state=None):
+            """LearningNodeAdaptive class constructor
+
+            Parameters
```
[the remark this answers is not in the document]
Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: The documentation was updated in the last commit.

File: src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py (hunk 1 above)
Comment (jacobmontiel) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: This change is indeed tricky for existing methods.

For SST-HT the change is straightforward. However, for existing methods, I think we can go with a deprecation warning for at least a couple of releases before removing the old names, see [here](https://stackoverflow.com/questions/9008444/how-to-warn-about-class-name-deprecation). I will create the issue to perform this change gradually, tentatively during the development of the next release.

Comment (smastelini) [reply src/skmultiflow/trees/stacked_single_target_regression_hoeffding_tree.py]: The stack overflow solution seems to be a good option on the long run. I will change the name of the class and send a new commit.

**Comment by jacobmontiel** [comment]
Code is correct and changes have been fully addressed. The method's correctness is granted from the fact that @smastelini is the author of SST-HT.

Issue #132 has been created to track any necessary updates from the paper's review.

PR is ready to merge. I will proceed with a squash commit due to the number of commits in the PR.

Thanks for the hard work @smastelini.

**Comment by smastelini** [comment]
Thank you @jacobmontiel for all the help during the PR process!

**Pull request merged**